File: view/lib/browser-policy.ts
const INTERNAL_DOMAIN_SUFFIXES = [
  '.local',
  '.internal',
  '.intranet',
  '.lan',
  '.corp',
  '.home',
  '.private',
  '.localdomain',
  '.localhost',
] as const

export const BROWSER_POLICY_HINT =
  '浏览器仅允许访问内网地址和内网域名（私有 IP、localhost、无点短名、*.local / *.internal / *.lan 等）'

function isCgnatV4(a: number, b: number): boolean {
  return a === 100 && (b & 0xc0) === 64
}

function parseIpv4(host: string): number[] | null {
  const m = host.match(/^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/)
  if (!m) return null
  const octets = m.slice(1).map(n => Number.parseInt(n, 10))
  if (octets.some(n => n > 255)) return null
  return octets
}

function isInternalIpv4(octets: number[]): boolean {
  const [a, b] = octets
  if (a === 10) return true
  if (a === 127) return true
  if (a === 0) return true
  if (a === 169 && b === 254) return true
  if (a === 192 && b === 168) return true
  if (a === 172 && b >= 16 && b <= 31) return true
  if (isCgnatV4(a, b)) return true
  return false
}

function isInternalIpv6Host(host: string): boolean | null {
  if (!host.includes(':')) return null
  const lower = host.toLowerCase()
  if (lower === '::1' || lower === '::') return true
  if (lower.startsWith('fc') || lower.startsWith('fd')) return true
  if (lower.startsWith('fe80')) return true
  return false
}

export function isInternalBrowserHost(host: string): boolean {
  const trimmed = host.trim().replace(/\.$/, '')
  if (!trimmed) return false

  const bare =
    trimmed.startsWith('[') && trimmed.endsWith(']') ? trimmed.slice(1, -1) : trimmed

  const ipv4 = parseIpv4(bare)
  if (ipv4) return isInternalIpv4(ipv4)

  const ipv6 = isInternalIpv6Host(bare)
  if (ipv6 !== null) return ipv6

  const lower = bare.toLowerCase()
  if (lower === 'localhost') return true

  if (!lower.includes('.')) return true

  return INTERNAL_DOMAIN_SUFFIXES.some(
    suffix => lower.endsWith(suffix) || lower === suffix.slice(1)
  )
}

export function assertInternalBrowserUrl(url: string): void {
  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    throw new Error('URL 无效')
  }
  const scheme = parsed.protocol.replace(':', '').toLowerCase()
  if (scheme !== 'http' && scheme !== 'https') {
    throw new Error('浏览器仅支持 http/https 协议')
  }
  if (!parsed.hostname) {
    throw new Error('URL 缺少主机名')
  }
  if (!isInternalBrowserHost(parsed.hostname)) {
    throw new Error(`${BROWSER_POLICY_HINT}，禁止访问: ${parsed.hostname}`)
  }
}

export function assertInternalBrowserHost(host: string): void {
  if (!isInternalBrowserHost(host)) {
    throw new Error(`${BROWSER_POLICY_HINT}，禁止访问: ${host}`)
  }
}

File: src-tauri/src/browser_policy.rs
use std::net::IpAddr;

const INTERNAL_DOMAIN_SUFFIXES: &[&str] = &[
    ".local",
    ".internal",
    ".intranet",
    ".lan",
    ".corp",
    ".home",
    ".private",
    ".localdomain",
    ".localhost",
];

/// 浏览器仅允许访问内网 IP 与内网域名。
pub fn validate_browser_url(url: &str) -> Result<(), String> {
    let parsed = tauri::Url::parse(url).map_err(|e| format!("URL 无效: {e}"))?;
    let scheme = parsed.scheme().to_ascii_lowercase();
    if scheme != "http" && scheme != "https" {
        return Err("浏览器仅支持 http/https 协议".to_string());
    }
    let host = parsed
        .host_str()
        .ok_or_else(|| "URL 缺少主机名".to_string())?;
    validate_browser_host(host)
}

pub fn validate_browser_host(host: &str) -> Result<(), String> {
    let host = host.trim().trim_end_matches('.');
    if host.is_empty() {
        return Err("主机名为空".to_string());
    }

    let bare = host
        .strip_prefix('[')
        .and_then(|h| h.strip_suffix(']'))
        .unwrap_or(host);

    if let Ok(ip) = bare.parse::<IpAddr>() {
        if is_internal_ip(ip) {
            return Ok(());
        }
        return Err(format!(
            "浏览器仅允许访问内网地址，禁止访问公网 IP: {host}"
        ));
    }

    let lower = bare.to_ascii_lowercase();
    if lower == "localhost" {
        return Ok(());
    }

    // 无点主机名视为内网短名（如 nginx、db01）
    if !lower.contains('.') {
        return Ok(());
    }

    for suffix in INTERNAL_DOMAIN_SUFFIXES {
        if lower.ends_with(suffix) || lower == suffix.trim_start_matches('.') {
            return Ok(());
        }
    }

    Err(format!(
        "浏览器仅允许访问内网地址和内网域名（如 *.local、*.internal、*.lan 或无点短名），禁止访问: {host}"
    ))
}

fn is_internal_ip(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            v4.is_private()
                || v4.is_loopback()
                || v4.is_link_local()
                || v4.is_unspecified()
                || is_cgnat_v4(v4)
        }
        IpAddr::V6(v6) => {
            v6.is_loopback()
                || v6.is_unspecified()
                || is_ula_v6(v6)
                || is_link_local_v6(v6)
        }
    }
}

fn is_cgnat_v4(ip: std::net::Ipv4Addr) -> bool {
    let o = ip.octets();
    o[0] == 100 && (o[1] & 0xC0) == 64
}

fn is_ula_v6(ip: std::net::Ipv6Addr) -> bool {
    (ip.segments()[0] & 0xfe00) == 0xfc00
}

fn is_link_local_v6(ip: std::net::Ipv6Addr) -> bool {
    (ip.segments()[0] & 0xffc0) == 0xfe80
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn allows_private_ipv4() {
        assert!(validate_browser_host("192.168.1.10").is_ok());
        assert!(validate_browser_host("10.0.0.5").is_ok());
        assert!(validate_browser_host("127.0.0.1").is_ok());
    }

    #[test]
    fn blocks_public_ipv4() {
        assert!(validate_browser_host("8.8.8.8").is_err());
        assert!(validate_browser_host("1.1.1.1").is_err());
    }

    #[test]
    fn allows_internal_domains() {
        assert!(validate_browser_host("svc.cluster.local").is_ok());
        assert!(validate_browser_host("app.internal").is_ok());
        assert!(validate_browser_host("nginx").is_ok());
    }

    #[test]
    fn blocks_public_domains() {
        assert!(validate_browser_host("google.com").is_err());
        assert!(validate_browser_host("baidu.com").is_err());
    }
}
